release: add release process doc (#16935)
snapcore/snapd: RELEASE.md

# Snapd Release Overview

This document describes the snapd release process and is primarily intended for snapd release engineers. We include the release process here publicly for both transparency and documentation purposes, though the process requires various administrative privileges unavailable to most.

*Note: At the beginning of every six-month cycle, in agreement with stakeholders, snapd creates a release plan that includes cut dates. Those cut dates determine what each release contains. If you are interested in getting particular content into a given release, be mindful of snapd's date-driven release process and check the release plan.*

> ## Release process

Reviewer: @maykathm, I created a diagram for the snapd special case documentation to try and capture a simplified high level overview of the release process. I'm wondering if perhaps we should bring that in, and try and use the phases as high level headings, and add more detail within as required for the detailed release process. 

The original: https://drive.google.com/file/d/1Rbs2l0YBaYQSxS4NgvzsoY8fL4gjBnu_/view
Here is a mermaid depiction: https://paste.ubuntu.com/p/3wgKBhxqps/

What do you think?

Reviewer: Perhaps its also useful to reference the special case documentation, because it contains lots of details about what is expected from who in terms of interactions with the SRU side of things.

It's currently WIP, but not expected to change significantly: https://github.com/ubuntu/ubuntu-project-docs/pull/542

Author: I've linked against that doc when under "Perform distro upgrade testing". Do you think it needs more of a mention?

Reviewer: I think thats fine for now, we still need to get the new version live.

The Snapd package deviates from the standard SRU process. The following special SRU process was followed: <link to relevant SRU process doc>

Release preparation: <link to release PR>
Release preparation test results: <link to test workflow results in GitHub>  
Failure analysis: <link to comment on PR explaining failures>

Release notes: <link to NEWS.md file from PR>

Please refer to snapd release notes documentation to understand how this is put together: https://docs.google.com/document/d/1do2TFwRIAzuOjLmteVuD0CRoJNO5vVcdUIwTHo4bYO4/edit?tab=t.0#heading=h.ablpjof536fg

Launchpad bugs addressed: https://launchpad.net/snapd/+milestone/<version>

Content overview:  <summary list of release content>

Areas of potential regressions: <only include if necessary>

Source packages on `ppa:snappy-dev/image` for upload to -proposed:
<list short name releases and the link for their deb packages>

Validation already completed:

- Release preparation test results: <link to release PR run results>
- QA Beta validation <link to beta validation Jira ticket>
- Certification validation: <link to test observer results> (need VPN)
```

> #### Examples of previous releases

Reviewer: This is nice.

# Snapd Release Overview

> This document describes the snapd release process and is primarily intended for snapd release engineers. We include the release process here publicly for both transparency and documentation purposes, though the process requires various administrative privileges unavailable to most.

Reviewer: team members?

Author: Do you mean you would like it to read "snapd release engineer team members" instead?

> *Note: At the beginning of every six-month cycle, in agreement with stakeholders, snapd creates a release plan that includes cut dates. Those cut dates determine what each release contains. If you are interested in getting particular content into a given release, be mindful of snapd's date-driven release process and check the release plan.*

Reviewer: We could mention "start of release" and "end of release" "estimates" , or something along those lines. I do not think cut dates alone works for planning.

Should we add comment to ourselves to link to release plan as something that we should start documenting publicly?

Author: I've changed the language and added a TODO to link to the release plan.

> *Note: At the beginning of every six-month cycle, in agreement with stakeholders, snapd creates a release plan that includes cut dates. Those cut dates determine what each release contains. If you are interested in getting particular content into a given release, be mindful of snapd's date-driven release process and check the release plan.*

Reviewer: Perhaps a heading ##Release planning?

## Release process

> The snapd release process consists of putting together content, crafting changelogs, building and releasing both snaps and debs along with cross-distro artifacts, and publicizing the release.

Reviewer: Perhaps link to further information.

(1) snapd snaps  <- link to snapd in snap store?
(2) Ubuntu debs <- link to launchpad snapd in ubuntu?
(3) other distros <- link to Maciej cross distro tracking page?
(4) Github tarballs <- link to releases in repo?

Author: Done

### Prerequisites

> The complete set of prerequisites for all release process steps, including optional ones, is as follows:

Reviewer: Distinguish optional for required?

- permission from snapd manager to release snapd snap to latest/beta
- a GPG key in both GitHub and Launchpad
- ability to create a release branch and push a tag to the snapd GitHub repository
- snappy-dev group membership in Launchpad
- permission to run autopkgtests in Launchpad (autopkgtest-requesters group membership)
- ability to promote snapd via snapcraft

> 

Reviewer: permission to re-trigger failing autopkgtests runnning on -proposed (optional)

Reviewer: I think its also useful to add link to information of where/how to get each of these.

Author: done and done

Besides listing the Ubuntu release targets in the description, they should also be selected as release targets. To select them, use the "Target to series" link and select the relevant releases. Once selected, they will appear in the table towards the top of the page.

Also subscribe relevant parties to the bug. In addition to key stakeholders, subscribe the following:
- [ubuntu-sru](https://launchpad.net/~ubuntu-sru)
- [sru-verification-interest-group-snapd](https://launchpad.net/~sru-verification-interest-group-snapd)

Base template:

```
New Snapd release <version> is required for <list of short name Ubuntu releases>.

Snapd <version> is the latest upstream release. It follows the previous Ubuntu release snapd <previous release version>.

The Snapd package deviates from the standard SRU process. The following special SRU process was followed: <link to relevant SRU process doc>

Release preparation: <link to release PR>
Release preparation test results: <link to test workflow results in GitHub>  
Failure analysis: <link to comment on PR explaining failures>

Release notes: <link to NEWS.md file from PR>

> Please refer to snapd release notes documentation to understand how this is put together: https://docs.google.com/document/d/1do2TFwRIAzuOjLmteVuD0CRoJNO5vVcdUIwTHo4bYO4/edit?tab=t.0#heading=h.ablpjof536fg

Reviewer 2: hm we really put a link to a non-public document in a public SRU bug report?

Author: Good point. @ernestl is this something we can change?

Reviewer: I think we need to make all the bits we need public

Reviewer 3: Indeed. There shouldn't be anything about the way we generate release notes that needs to be kept internal, although I recognise the increase in scope here.

Author: I've added what I think are the important bits from that documentation to this release doc. I then changed the template to point back to this doc. Let me know if this makes sense.

3. Update NEWS.md with the list created in the previous section
4. Generate changelogs using the NEWS.md created in the previous step
	- Run the changelog script as follows: `DEBEMAIL="<name> <email>" release-tools/changelog.py <version> <sru bug number> NEWS.md` 
		- e.g. `DEBEMAIL="Katie May <katie.may@canonical.com>" release-tools/changelog.py 2.75.2 2143882 NEWS.md`
	- Do not commit the markdown file generated called `snapd-<version>-GitHub-release.md`. It can be used to help create the GitHub release, if so desired. Otherwise, it can be ignored.
	- Double-check changed files
		- `git diff | diffstat`
		- `git diff release/2.75 | diffstat` (any recent previous tag will do)
		- the same files should get touched, all inside the packaging/tree
	- If you are generating changelogs for a version that supersedes the major version, remove `New upstream releases, LP#<lp-number>` from the previous version. The `+<ubuntu release>` and the release name from the previous version should match those of the current minor version.
4. Commit NEWS.md and the changelog updates as `release: <version>` into your personal branch
5. Open the PR against `release/<version>`

> 	- The PR description should be formatted as follows (only include the cherry picked section if you have cherry-picked commits):

Reviewer: Should we perhaps create a PR template in github and refer to it?

Author: I looked into creating a PR template. It looks to me like managing multiple templates is not amazing and might confuse people. I've left it as is. Feel free to push back if you think we should go ahead with it.

- If the original reporter is external to the organization, encourage them to help verify the fix.
- If the original reporter is internal to the organization, contact them and ask for testing support.

### 8. Request upload to archive release pocket

## Cross-distro releases

Prerequisites
- snapd snap is in candidate and the GitHub release is public

### Opensuse

#### One-time setup
- Create an account on build.opensuse.org and ask an existing project admin to add you to system:snappy.
- On your openSUSE machine, install the osc tool, then checkout the snapd package from OBS: `osc checkout system:snappy`
- We then need to install various packages needed for building RPMs: `sudo zypper install rpmdevtools`

#### Making a new release

Steps:
1. Branch and checkout in local dir (`bco`)
2. Retrieve tarballs, dropping the old one and adding the new one
3. Edit the spec
4. Prep tree for `rpmbuild`
5. Copy sources
6. Build the binary
7. Install, check the versions, run some snaps like `ohmygiraffe` that uses `opengl` and `lxd`
8. Edit the changelog (`osc vc`)
9. Look at the diff

> 10. Check in and submit the package (`osc ci` or through the webpage)

Reviewer 2: ```suggestion
10. Check in and submit the package (`osc ci` or through the webpage)
11. Enable publishing of RPMs in your OBS branch
12. Run the opensuse workflow in https://github.com/canonical/snapd-smoke-tests/ pointing it to your fork
```

> ```

Reviewer 2: I think I'd drop the actual commands. They will go out of date.

osc bco system:snappy snapd
cd home:<obs-user>:branches:system:snappy
cd snapd
wget https://github.com/snapcore/snapd/releases/download/2.56.2/snapd_2.56.2.vendor.tar.xz
osc rm snapd_2.56.vendor.tar.xz
osc add snapd_2.56.2.vendor.tar.xz

> vim snapd.spec

Reviewer 2: this is what I usually do:
```suggestion
vim snapd.spec    # bump the version inside
osc service runall
```

rpmdev-wipetree
rpmdev-setuptree
cp snapd_2.56.2.vendor.tar.xz ~/rpmbuild/SOURCES
cp snapd-rpmlintrc ~/rpmbuild/SOURCES
rpmbuild -ba ./snapd.spec  # use zypper to install any missing dep
sudo zypper in ~/rpmbuild/RPMS/x86_64/snapd-2.56.2-0.x86_64.rpm
snap version
osc vc
osc diff
osc st
osc ci
osc submitpac
```

### Archlinux

#### One-time setup
- Create an account on the Arch User Repository (AUR) at aur.archlinux.org, then ask an existing maintainer to add you to the snapd project.

> - On your Arch machine, install the following packages: `pacman -S base-devel git asciidoc pkg-config moreutils`

Reviewer 2: ```suggestion
- On your Arch machine, install the following packages: `pacman -S base-devel git asciidoc pkg-config moreutils aurpublish`
```

- Checkout the repositories:
	- `git clone ssh://aur@aur.archlinux.org/snapd.git`
	- `git clone https://github.com/eli-schwartz/aurpublish.git`
- Build the `aurpublish` tool:
	- `cd aurpublish`
	- `make && sudo make install`
- Verify that you can build snapd:

> 	- `cd ../snapd`

Reviewer 2: ```suggestion
- Verify that you can build snapd:
	- `cd snapd`
```

	- `makepkg -s`
- Download this script, which will be used to update the snapd version number and checksums in the PKGBUILD file: https://gist.github.com/bboozzoo/6780ebeaac781c9ee57c0ec24252a55e
- Save it as `upgpkg`, make it executable and put it somewhere in your `$PATH`.

#### Making a new release
Enter the directory where the snapd PKGBUILD file is located, and run
`upgpkg 2.56.3  # Write the new snapd version number`

> 

Reviewer 2: ```
Alternatively fork https://github.com/bboozzoo/pkgbuilds, submit PR, run `arch`
workflow in https://github.com/canonical/snapd-smoke-tests/ providing it with a link to the PR.
```

    classDef qa fill:#FDCFB7,stroke:#E95420,color:#000
    classDef cert fill:#E0C8DE,stroke:#772953,color:#000
    classDef sru fill:#B8CCE4,stroke:#1a4a7a,color:#000
    classDef external fill:#B8DEBB,stroke:#2a7a2a,color:#000
    classDef label fill:#3a3a3a,stroke:#555,color:#ddd

    style t0 fill:#2a2a2a,stroke:#3a3a3a,color:#ddd,font-size:16px
    style t1 fill:#2a2a2a,stroke:#3a3a3a,color:#ddd,font-size:16px
    style t2 fill:#2a2a2a,stroke:#3a3a3a,color:#ddd,font-size:16px
    style t3 fill:#2a2a2a,stroke:#3a3a3a,color:#ddd,font-size:16px
    style t4 fill:#2a2a2a,stroke:#3a3a3a,color:#ddd,font-size:16px
    style w0 fill:#222,stroke:#333
    style w1 fill:#222,stroke:#333
    style w2 fill:#222,stroke:#333
    style w3 fill:#222,stroke:#333
    style w4 fill:#222,stroke:#333

    subgraph t0 [" "]
        T0L["<b>Prepare release content - 1 day</b>"]:::label
        subgraph w0 [" "]
            D1{{"Agree on release content"}}:::decision
        end
    end

    D1 --> RT["Test release"]:::qa

    subgraph t1 [" "]
        T1L["<b>Build beta artifacts - 1 week</b>"]:::label
        subgraph w1 [" "]
            RT
            BS["Build snaps"]:::qa

>             BD["Build deps"]:::sru --> QS["QA testing<br/>(snappy-dev/image)"]:::qa

Reviewer: debs  or Debian packages

- GitHub release https://github.com/canonical/snapd/releases
- Cross-distro artifacts https://snapcraft.io/docs/reference/administration/distribution-support/

*Note: documentation is produced separately from the [snap-docs repository](github.com/canonical/snap-docs)*

### Prerequisites

The complete set of prerequisites for all release process steps is as follows:
- permission from snapd manager to release snapd snap to latest/beta
- a GPG key in both [GitHub](https://docs.github.com/en/authentication/managing-commit-signature-verification/adding-a-gpg-key-to-your-github-account) and [Launchpad](https://documentation.ubuntu.com/launchpad/user/how-to/import-openpgp-key/)
- ability to create a release branch and push a tag to the snapd GitHub repository
- snappy-dev group membership in Launchpad
- ability to promote snapd via snapcraft (you can check if you have permission by running `snapcraft status snapd`)
- permission to run autopkgtests in Launchpad (autopkgtest-requesters group membership; request via debcrafters)
- permission to re-trigger failing autopkgtests running on -proposed (request via debcrafters)

# Full Release Process

## Prepare release content

> **Time: 1 day**

Reviewer: Opinion: Not that important, but in my experience this is very optimistic.

Author: Changed to a week

- release branch creation rights

Create a branch named `release/<version-number>` from the master branch on [github.com/canonical/snapd](https://github.com/canonical/snapd). If you are creating a patch release, do not create a new branch; use the corresponding `release/2.XX` branch instead.

#### 2. Create a new SRU tracking bug in Launchpad

Create the bug under https://bugs.launchpad.net/ubuntu/+source/snapd (and not https://bugs.launchpad.net/snapd)

It should be named `[SRU] <version>`

NOTE: As a general rule, all releases require their own new SRU bug. The exception is patch releases that are intended to supersede a previous release that will not be pursued any further.

Besides listing the Ubuntu release targets in the description, also select them as release targets. To do that, use the "Target to series" link and select the relevant releases. Once selected, they will appear in the table near the top of the page.

Also subscribe relevant parties to the bug. In addition to key stakeholders, subscribe the following:
- [ubuntu-sru](https://launchpad.net/~ubuntu-sru)
- [sru-verification-interest-group-snapd](https://launchpad.net/~sru-verification-interest-group-snapd)

> Base template:

Reviewer: Once we get the special case docs approved I think we can reference that for the template.
I would call this `Launchpad SRU Bug description template` or slightly shortened version of that.

Part of the reason is, we also have a different thing `Launchpad test templates` for all the LP bugs associated with our release, and this is unusual from SRU point of view (its in our exception) and worth clearly differentiating.

The Snapd package deviates from the standard SRU process. The following special SRU process was followed: <link to relevant SRU process doc>

Release preparation: <link to release PR>
Release preparation test results: <link to test workflow results in GitHub>  
Failure analysis: <link to comment on PR explaining failures>

Release notes: <link to NEWS.md file from PR>

Please refer to snapd release notes documentation (https://github.com/canonical/snapd/blob/master/RELEASE.md) for details on how this is assembled.

Launchpad bugs addressed: https://launchpad.net/snapd/+milestone/<version>

Content overview:  <summary list of release content>

Areas of potential regressions: <only include if necessary>

Source packages on `ppa:snappy-dev/image` for upload to -proposed:
<list short name releases and the link for their deb packages>

Validation already completed:

- Release preparation test results: <link to release PR run results>
- QA Beta validation <link to beta validation Jira ticket>
- Certification validation: <link to test observer results> (need VPN)

Verification required:

> - Verify LP bugs for <list of short name target releases>

Reviewer: I think we need to provide a sample test template. And explain its relationship to the SRU LP Bug.

Author: added a new step for adding SRU test templates

- Perform release upgrades from <short name of earliest supported target release> to <short name of latest supported release>
```

> ##### Notable examples illustrating selected edge cases and fully completed releases:

Reviewer: This is great!

- [2.74.1 - supersede example](https://bugs.launchpad.net/ubuntu/+source/snapd/+bug/2138629)
- [2.73](https://bugs.launchpad.net/ubuntu/+source/snapd/+bug/2132084)
- [2.70 - unreleased](https://bugs.launchpad.net/ubuntu/+source/snapd/+bug/2112209)
- [2.68](https://bugs.launchpad.net/ubuntu/+source/snapd/+bug/2098137)
- [2.65.1 - Feature freeze exception](https://bugs.launchpad.net/ubuntu/+source/snapd/+bug/2078050)
- [2.61.3 - patch release example](https://bugs.launchpad.net/ubuntu/+source/snapd/+bug/2039017)

#### 3. Curate list of changes for NEWS.md

NEWS.md contains the list of functional changes in each snapd release for snapd developers. The release engineer should carefully review all release content, including anything that will be cherry-picked, and create a clear, comprehensive list of the functional changes included in the release.

> NEWS.md properties:

Reviewer: (1) Should we perhaps add that PRs related to the same theme or area should share the same `<prefix>:` and be grouped together. 

(2) All fixes related to LP bugs - that impact Ubuntu Classic / Hybrid Classic - should have `LP: #<NUMBER>` before the note. These are automatically tracked by launchpad.

(3) Interface convention?

Author: added

- it should contain a full, linear, historical record of candidate releases

> - it should not include tests or non-functional changes

Reviewer: Nitpick: Also not "internal fixes" - something that was both broken and fixed/reverted between two release points.

Author: changed

2. If there are commits to include after the release was cut, cherry pick any needed commits:
	- All cherry picked commits should have the "cherry-picked" label added to their merged PR and ensure their milestone is properly set
	- One can update [this query](https://github.com/snapcore/snapd/pulls?q=-label%3Acherry-picked+is%3Apr+milestone%3A2.61+sort%3Aupdated-desc) to see which PRs target the release but have not yet been cherry-picked.
3. Update NEWS.md with the list created in the previous section
4. Generate changelogs using the NEWS.md created in the previous step
	- Run the changelog script as follows: `DEBEMAIL="<name> <email>" release-tools/changelog.py <version> <sru bug number> NEWS.md`
    - Do not commit the generated markdown file `snapd-<version>-GitHub-release.md`. You can use it to help create the GitHub release if needed; otherwise, ignore it.
	- Double-check changed files
		- `git diff | diffstat`
		- `git diff release/2.75 | diffstat` (any recent previous tag will do)
        - The same files should get touched, all within `packaging/tree`.

> 	- If you are generating changelogs for a version that supersedes a previous version, remove `New upstream releases, LP#<lp-number>` from the previous version. The `+<ubuntu release>` and the release name from the previous version should match those of the current patch version.

Reviewer: This is the convention: `LP: #2118396`

**IMPORTANT: the tag must be signed with your GPG key uploaded both to GitHub and Launchpad**

1. Pull the release branch with the new commits after the rebase merge
2. `git tag -s 2.XX -m "tagging package snapd version 2.XX"`
3. `git push upstream 2.XX` (assuming canonical/snapd is upstream)

#### 3. Merge changelogs back into master

1. Create a new branch in your fork called `changelogs-<version>` from the tag you just pushed (2.XX)
2. Merge master into the `changelogs-<version>` branch with the `--no-ff` option. If there are conflicts, always choose the changes from master: `git merge --no-ff master`
3. Open a PR to merge into master (be sure to skip spread tests) with the description: `Merge changelogs back into master, please use a regular “merge” to merge it.`
	1. The PR should contain the complete merge history, including the tagged commit
	2. You should use a regular merge commit to merge it (not squash and merge, not rebase and merge)
4. Once merged, the tag's commit should be found on the master branch `git branch --contains <tag/tag-commit>`

#### 4. Create snapd snap builds for `beta/<version>` on Launchpad

> **IMPORTANT: Only trigger the snapd snap builds once you see the tag has been imported to Launchpad. The version is derived from the tag, so its absence will incorrectly produce `<version>+git`**

Reviewer: Love the helpful hints!

1. Go to https://code.launchpad.net/~snappy-dev/snapd/+git/snapd
2. Click "import now" to import the codebase
3. Click on https://code.launchpad.net/~snappy-dev/snapd/+git/snapd/+ref/release/2.XX and attempt to browse the code. It will probably not work, but if it does work, then you can confirm the presence of the git tag.
4. Go back to the `release/2.XX` branch (https://code.launchpad.net/~snappy-dev/snapd/+git/snapd/+ref/release/2.XX) and click "Create snap package" (https://code.launchpad.net/~snappy-dev/snapd/+git/snapd/+ref/release/2.XX/+new-snap).

> 5. In a different window, open a [previous package in edit mode](https://launchpad.net/~snappy-dev/+snap/snapd-2.74/+edit), and copy that configuration to the newly created package. Change the recipe and branch name.

Reviewer 2: ```suggestion
5. In a different window, open a [previous package in edit mode](https://launchpad.net/~snappy-dev/+snap/snapd-2.74/+edit), and copy that configuration to the newly created package. Change the recipe name to `snapd-2.XX` and branch name to `release/2.XX`.
```

2. Contact the assigned tester from cert, clarify timeline and priority, and ask them to contact us immediately for any unexpected issues.
3. Update https://forum.snapcraft.io/t/the-snapd-roadmap/1973 with release notes content, the correct date, and a green checkbox for beta.
4. Once snapd beta testing and certification testing have concluded, sync with snapd QA and promote to candidate.

#### 7. Build snapd debs and tarballs

*Note: It may sometimes be necessary to add content to the debs that is not in the snap (e.g. to fix an autopkgtest). The preferred way of doing so is to open a PR against the release branch with the cherry-picked changes.*

##### Preliminary work

If you have never done this step before, you might need to patch your `gbp` tool:
```
(cd /usr/lib/python3/dist-packages/gbp ; curl 'https://bugs.debian.org/cgi-bin/bugreport.cgi?att=1;bug=894790;filename=gbp-support-symlinks.diff;msg=5' | sudo patch -p2 )
```

If you have never installed snapd build dependencies, after creating the debian directory in the steps below, run
```
sudo apt build-dep .
```

##### Steps
 
1. Check out the `release/<version>` branch

> 2. Since the `debian` symlink was removed from the source, create the `debian` folder symlink: `ln -sfn packaging/ubuntu-16.04 debian`

Reviewer: Unfortunately, we have to, for the moment also do this step for packaging/ubuntu-26.04 which is not a thing since we dropped transitional packages and bumped gh_helper level, but its essentially a repeat with just symlink to different folder and repeat. I do not think we should say a lot about this, we need improve around this with helper script etc, but for the moment it worth knowing.

Author: changed

# Snapd Release Overview

This document describes the snapd release process and is primarily intended for release engineers in the snapd team. We include the release process here publicly for both transparency and documentation purposes, though the process requires various administrative privileges unavailable to most.

## Release planning

At the beginning of every six-month cycle, in agreement with stakeholders, snapd creates a release plan that includes estimated start and end dates for releases. Those dates determine what each release contains. If you want particular content to land in a given release, be mindful of snapd's date-driven release process and check the release plan.

*TODO: add a link to the release plan*

## Release process

> The snapd release process consists of putting together content, crafting changelogs, testing, building and releasing both snaps and debs along with cross-distro artifacts, and publicizing the release.

Reviewer 3: You should divide this into the same phases as on the diagram, maybe as a bulleted list.

Author: I removed them all together. Since the diagram immediately follows, I don't think we get any added value from listing the phases.

### Overview

```mermaid
flowchart TD
    classDef decision fill:#F5D699,stroke:#D4A017,color:#000
    classDef qa fill:#FDCFB7,stroke:#E95420,color:#000
    classDef cert fill:#E0C8DE,stroke:#772953,color:#000
    classDef sru fill:#B8CCE4,stroke:#1a4a7a,color:#000
    classDef external fill:#B8DEBB,stroke:#2a7a2a,color:#000
    classDef label fill:#3a3a3a,stroke:#555,color:#ddd

    style t0 fill:#2a2a2a,stroke:#3a3a3a,color:#ddd,font-size:16px
    style t1 fill:#2a2a2a,stroke:#3a3a3a,color:#ddd,font-size:16px
    style t2 fill:#2a2a2a,stroke:#3a3a3a,color:#ddd,font-size:16px
    style t3 fill:#2a2a2a,stroke:#3a3a3a,color:#ddd,font-size:16px
    style t4 fill:#2a2a2a,stroke:#3a3a3a,color:#ddd,font-size:16px
    style w0 fill:#222,stroke:#333
    style w1 fill:#222,stroke:#333
    style w2 fill:#222,stroke:#333
    style w3 fill:#222,stroke:#333
    style w4 fill:#222,stroke:#333

    subgraph t0 [" "]

>         T0L["<b>Prepare release content - 6 weeks</b>"]:::label

Reviewer 3: Would prefer that the release process initiates at the point where we approve the content, rather than 6 weeks earlier.

i.e. "PR 1/2/N testing" doesn't in fact happen under the responsibility of the release engineer.

Author: Changed. @ernestl I changed it to being 1 day to cut the release and create the PR. Sound good?

        subgraph w0 [" "]
            PR1["PR 1 testing"]:::qa
            PR2["PR 2 testing"]:::qa
            PRN["PR N testing"]:::qa

>             PR1 & PR2 & PRN --> D1{{"Review release content"}}:::decision

Reviewer 3: "Agree" rather than review

        end
    end

>     D1 --> RT["Release testing"]:::qa

Reviewer 3: "Release testing" in here feels like the title of the block as opposed to an actual step

Author: Changed to "Test release"

    subgraph t1 [" "]

>         T1L["<b>Build and sanity check artifacts - 1 week</b>"]:::label

Reviewer 3: "beta artifacts"

        end
    end
    RT --> BS & BD

    subgraph t2 [" "]
        T2L["<b>Beta testing - 1 week</b>"]:::label
        subgraph w2 [" "]
            QB["QA testing (beta)"]:::qa
            PB["Public testing (beta)"]:::external
            CB["Certification testing (beta)"]:::cert
            DA["Dev autopkgtests<br/>(-proposed)"]:::sru --> IA["Interim/LTS autopkgtests<br/>(-proposed)"]:::sru --> QP["QA testing<br/>(-proposed)"]:::qa
            PD["Public Dev testing<br/>(-proposed)"]:::external
        end
    end
    BS --> PB & CB & QB
    QS --> DA & PD

    PB & CB & QB & QP & PD --> D2{{"Review known issues"}}:::decision

    subgraph t3 [" "]
        T3L["<b>Candidate testing - 1 week</b>"]:::label
        subgraph w3 [" "]
            D2
            PC["Public testing (candidate)"]:::external
            IT["Public Interim/LTS testing<br/>(-proposed)"]:::sru
            DR["Dev release deb<br/>(-updates)"]:::sru
            XD["Start cross distro<br/>release process"]:::sru
        end
    end
    D2 --> PC
    D2 --> IT & DR
    D2 --> XD

    subgraph t4 [" "]
        T4L["<b>Release - 1 week</b>"]:::label
        subgraph w4 [" "]

>             SS["Stable release (snap)"]:::external

Reviewer 3: add "progressive"

            IR["Interim/LTS releases deb<br/>(-updates)"]:::sru
        end
    end
    PC --> SS
    IT & DR --> IR
```

```mermaid
%%{init: {'theme': 'dark', 'themeVariables': {'fontSize': '11px'}, 'flowchart': {'padding': 4, 'nodeSpacing': 10, 'rankSpacing': 10}}}%%
flowchart LR
    classDef qa fill:#FDCFB7,stroke:#E95420,color:#000
    classDef cert fill:#E0C8DE,stroke:#772953,color:#000
    classDef sru fill:#B8CCE4,stroke:#1a4a7a,color:#000
    classDef external fill:#B8DEBB,stroke:#2a7a2a,color:#000
    classDef decision fill:#F5D699,stroke:#D4A017,color:#000

    L1["Snapd Team"]:::qa ~~~ L2["Certification Team"]:::cert ~~~ L3["Ubuntu SRU & Snapd Teams"]:::sru ~~~ L4["Canonical & External Stakeholders"]:::external ~~~ L5["Snapd Team Decision"]:::decision
```

### Release output artifacts

The release process produces the following:
- The snapd snap https://snapcraft.io/snapd
- snapd debs https://launchpad.net/ubuntu/+source/snapd/
- GitHub release https://github.com/canonical/snapd/releases
- Cross-distro artifacts https://snapcraft.io/docs/reference/administration/distribution-support/

> 

Reviewer 3: "Note: documentation is produced separately from the snap-docs repository"

### Prerequisites

The complete set of prerequisites for all release process steps, including optional ones, is as follows:
- permission from snapd manager to release snapd snap to latest/beta
- a GPG key in both [GitHub](https://docs.github.com/en/authentication/managing-commit-signature-verification/adding-a-gpg-key-to-your-github-account) and [Launchpad](https://documentation.ubuntu.com/launchpad/user/how-to/import-openpgp-key/)
- ability to create a release branch and push a tag to the snapd GitHub repository
- snappy-dev group membership in Launchpad
- ability to promote snapd via snapcraft (you can check if you have permission by running `snapcraft status snapd`)
- (optional) permission to run autopkgtests in Launchpad (autopkgtest-requesters group membership; request via debcrafters)
- (optional) permission to re-trigger failing autopkgtests running on -proposed (request via debcrafters)

# Full Release Process

## Prepare release content

**Time: 6 weeks**

### Summary

> During the content preparation phase, PRs are created, tagged with the relevant milestone, and merged to master. As the cutoff approaches, the team should regularly check the GitHub milestone and work together to merge the content targeted for the release. One week before the cut date, the team should ensure any broken tests on master are diagnosed and fixed. On the cut date, the release engineer creates a new release branch and opens a PR against that branch with changelogs covering all relevant release content. CI testing on that PR kicks off the next phase: building and sanity-checking artifacts.

Reviewer 3: Same as above, preparing content isn't in the scope of the release process itself. It begins at the point we approve the cut.

### Steps

#### 1. Create release branch

Prerequisites
- release branch creation rights

> Create a branch named `release/<version-number>` from the master branch on [github.com/canonical/snapd](https://github.com/canonical/snapd). If you are creating a dot release, do not create a new branch; use the corresponding `release/<major>.<minor>` branch instead.

Reviewer 3: Probably one of the more annoying things (for you!) that I've noticed whilst reviewing this.

In many cases in this doc, we are using major/minor in a contradictory way.

The "2" is the major version number. But if you look lower down, in many places it's treating the number after the 2 as a "major" release.

Also, what we call a "dot" release, semver calls a "patch" release.

But then, we have a misalignment with SSDLC, because that has the concept of a "major" release every LTS cycle.

This probably needs some thought, and a longer conversation.

Author: Yes, quite right. I updated it 's/dot/patch/g' and avoided major/minor in attempt to not confuse.

#### 2. Create a new SRU tracking bug in Launchpad

Create the bug under https://bugs.launchpad.net/ubuntu/+source/snapd (and not https://bugs.launchpad.net/snapd)

It should be named `[SRU] <version>`

> NOTE: All releases require their own new SRU bug, except dot releases that are intended to supersede a previous release, whether major or dot, that will not be released.

Reviewer 3: patch?

> NOTE: All releases require their own new SRU bug, except dot releases that are intended to supersede a previous release, whether major or dot, that will not be released.

Reviewer 2: Found this a bit hard to follow. How about:
```suggestion
NOTE: As a general rule, all releases require their own new SRU bug. The exception is dot releases that are intended to supersede a previous release, major or dot, that will not be pursued any further.
```

or replacing `,major or dot,` with `(major or dot)`

Release preparation: <link to release PR>
Release preparation test results: <link to test workflow results in GitHub>  
Failure analysis: <link to comment on PR explaining failures>

Release notes: <link to NEWS.md file from PR>

Please refer to snapd release notes documentation (https://github.com/canonical/snapd/blob/master/RELEASE.md) for details on how this is assembled.

Launchpad bugs addressed: https://launchpad.net/snapd/+milestone/<version>

Content overview:  <summary list of release content>

Areas of potential regressions: <only include if necessary>

Source packages on `ppa:snappy-dev/image` for upload to -proposed:
<list short name releases and the link for their deb packages>

Validation already completed:

- Release preparation test results: <link to release PR run results>
- QA Beta validation <link to beta validation Jira ticket>
- Certification validation: <link to test observer results> (need VPN)

Verification required:

- Verify LP bugs for <list of short name target releases>
- Perform release upgrades from <short name of earliest supported target release> to <short name of latest supported release>
```

> ##### Examples of previous releases

Reviewer 3: I think a couple of examples would be fine.

- [2.75.2](https://bugs.launchpad.net/ubuntu/+source/snapd/+bug/2143882)
- [2.74.1 - supersede example](https://bugs.launchpad.net/ubuntu/+source/snapd/+bug/2138629)
- [2.73](https://bugs.launchpad.net/ubuntu/+source/snapd/+bug/2132084)
- [2.72](https://bugs.launchpad.net/ubuntu/+source/snapd/+bug/2124239)
- [2.71](https://bugs.launchpad.net/ubuntu/+source/snapd/+bug/2118396)
- [2.70](https://bugs.launchpad.net/ubuntu/+source/snapd/+bug/2112209)
- [2.69](https://bugs.launchpad.net/ubuntu/+source/snapd/+bug/2105854)
- [2.68](https://bugs.launchpad.net/ubuntu/+source/snapd/+bug/2098137)
- [2.67](https://bugs.launchpad.net/ubuntu/+source/snapd/+bug/2089691)
- [2.66](https://bugs.launchpad.net/ubuntu/+source/snapd/+bug/2083490)
- [2.65.1 - Feature freeze exception](https://bugs.launchpad.net/ubuntu/+source/snapd/+bug/2078050)
- [2.65](https://bugs.launchpad.net/ubuntu/+source/snapd/+bug/2077473)
- [2.64](https://bugs.launchpad.net/ubuntu/+source/snapd/+bug/2072986)
- [2.63](https://bugs.launchpad.net/ubuntu/+source/snapd/+bug/2061179)
- [2.62](https://bugs.launchpad.net/ubuntu/+source/snapd/+bug/2058277)

> - [2.61.3 - dot release example](https://bugs.launchpad.net/ubuntu/+source/snapd/+bug/2039017)

Reviewer 2: ```suggestion
Notable examples illustrating selected edge cases and fully completed releases:

- [2.75.2](https://bugs.launchpad.net/ubuntu/+source/snapd/+bug/2143882)
- [2.74.1 - Supersede example](https://bugs.launchpad.net/ubuntu/+source/snapd/+bug/2138629)
- [2.70 - Unreleased](https://bugs.launchpad.net/ubuntu/+source/snapd/+bug/2112209)
- [2.68](https://bugs.launchpad.net/ubuntu/+source/snapd/+bug/2098137)
- [2.65.1 - Feature freeze exception](https://bugs.launchpad.net/ubuntu/+source/snapd/+bug/2078050)
- [2.61.3 - Dot release example](https://bugs.launchpad.net/ubuntu/+source/snapd/+bug/2039017)
```

#### 3. Curate list of changes for NEWS.md

NEWS.md contains the list of functional changes in each snapd release for snapd developers. The release engineer should carefully review all release content, including anything that will be cherry-picked, and create a clear, comprehensive list of the functional changes included in the release.

NEWS.md properties:
- it should contain a full, linear, historical record of candidate releases
- it should not include tests or non-functional changes
- it should cover each externally relevant PR
- comments should be concise and understandable
- summarized comments should be grouped in an order that is helpful to the audience, e.g. new features, bugfixes, and interface changes grouped together

Steps:

> - (Optional) Use a script to gather all PRs in the release and create a spreadsheet to work from.

Reviewer 3: is this script checked in anywhere?

Author: I added it inline

#### 4. Create release PR

Prerequisites
- Python packages: python-debian, bs4, markdown

The PR should be from a branch created in your own snapd fork against the upstream release branch created in an earlier step. The branch must include at least an updated NEWS.md and changelogs and might additionally contain cherry-picked commits.

Steps:
1. Create a branch on your own snapd fork from the upstream `release/<version>` branch and call it `release-<version>` (e.g. `release-2.61`)
2. If there are commits to include after the release was cut, cherry pick any needed commits:
	- All cherry picked commits should have the "cherry-picked" label added to their merged PR and ensure their milestone is properly set
	- One can update [this query](https://github.com/snapcore/snapd/pulls?q=-label%3Acherry-picked+is%3Apr+milestone%3A2.61+sort%3Aupdated-desc) to see which PRs target the release but have not yet been cherry-picked.
3. Update NEWS.md with the list created in the previous section
4. Generate changelogs using the NEWS.md created in the previous step
	- Run the changelog script as follows: `DEBEMAIL="<name> <email>" release-tools/changelog.py <version> <sru bug number> NEWS.md` 

> 		- e.g. `DEBEMAIL="Katie May <katie.may@canonical.com>" release-tools/changelog.py 2.75.2 2143882 NEWS.md`

Reviewer 3: please remove your name from the documented process

		
		LP bugs: https://launchpad.net/snapd/+milestone/<version>
		SRU bug: https://bugs.launchpad.net/snapd/+bug/<sru bug number>
		Jira: <link to jira ticket>
		
		CHERRY PICKED
		Functional fixes:
		- list of cherry-picked functional fixes
		Test fixes:
		- list of cherry-picked test fixes
		
		Requires rebase merge
		```
7. Add the "Run nested" label, then add the senior engineer and all team members who have content in NEWS.md as reviewers.

## Build and sanity-check artifacts

**Time: 1 week**

### Summary

The phase begins by testing the PR created against the release branch at the end of the previous phase. Once tests have either passed or their failure been proven irrelevant, the PR gets merged. After pushing a version tag to that branch, the release engineer then opens and merges a PR against master with the updated changelogs and the git tag. With a release version tag now visible on master, the release engineer builds the snapd snap on Launchpad and the debs locally. Those artifacts are then tested: the snap via snapd beta testing and the debs via autopkgtests in snappy-dev.

### Steps

#### 1. Test release PR and merge it

> 1. Once the PR is approved, rerun tests as many times as necessary to get past flaky failures. 

Reviewer 3: s/flaky/intermittent/

Push the version tag to the canonical/snapd repo by following the steps:

**IMPORTANT: the tag must be signed with your GPG key uploaded both to GitHub and Launchpad**

1. Pull the release branch with the new commits after the rebase merge
2. `git tag -s 2.XX -m "tagging package snapd version 2.XX"`
3. `git push upstream 2.XX` (assuming canonical/snapd is upstream)

#### 3. Merge changelogs back into master

1. Create a new branch in your fork called `changelogs-<version>` from the tag you just pushed (2.XX)
2. Merge master into the `changelogs-<version>` branch with the `--no-ff` option. If there are conflicts, always choose the changes from master: `git merge --no-ff master`
3. Open a PR to merge into master (be sure to skip spread tests) with the description: `Merge changelogs back into master, please use a regular “merge” to merge it.`
	1. The PR should contain the complete merge history, including the tagged commit
	2. You should use a regular merge commit to merge it (not squash and merge, not rebase and merge)
4. Once merged, the tag's commit should be found on the master branch `git branch --contains <tag/tag-commit>`

#### 4. Garden the GitHub milestone

> Ensure that only future milestones remain, or that there is a single open milestone.

Reviewer 3: I think we can remove this bit

6. Once you are done setting it up, save, click on the package (https://launchpad.net/~snappy-dev/+snap/snapd-2.XX), and request builds.
7. Once the builds have completed, ensure the versions are correct by checking `snapcraft status snapd | grep beta/<version>`

#### 6. Release to latest/beta

Prerequisites:
- You have the necessary snapcraft permissions to promote the snapd snap
- You have permission from the snapd manager to promote to latest/beta
- Builds on all architectures have succeeded

Steps:
1. Sync with the QA person on the team in charge of beta testing to make sure we can go to beta. For example, if a previous release has not yet made it to candidate, we would need to hold off on promoting.
2. Find the revisions for the snap you just pushed. It will be `beta/<version>`. You can find the revisions by running `snapcraft status snapd`
3. For each architectural build, you have a unique revision number. You release by running `snapcraft release snapd <revision-number> beta`. You will run that command for each individual revision number.

> 

Reviewer 3: Could you add "Update internal roadmap tracking, for example by marking Jira epics and releases as completed"

And also put the "Update GitHub milestones to close the released milestone"

	1. Check for the details of the most recent deb release https://launchpad.net/~snappy-dev/+archive/ubuntu/image/+packages?field.name_filter=snapd&field.status_filter=&field.series_filter=
	2. In `debian/changelog` change that entry to reflect the correct version (e.g. `snapd (2.75.1) xenial` -> `snapd (2.75.1+ubuntu26.04) resolute` when building for resolute). If the latest deb package on the PPA is not present in the changelog, add it at the appropriate place. Also change all later changelog entries to reflect the correct version.
		- *NOTE: If you did a re-spin and are releasing multiple versions, remove the `New upstream release` line in the version that never got released*
	3. Build the package from the directory `build-area/snapd-<version>` (`dpkg-buildpackage -S`)
6. Create the source tarballs from the `build-area` directory: `../snapd/release-tools/repack-debian-tarball.sh ./snapd_<version>.tar.xz`

#### 9. Upload tarballs to GitHub release

From the previous step, the `repack-debian-tarball.sh` script will have created the following files:
- `snapd_<version>.no-vendor.tar.xz`

> -  `snapd_<version>.only-vendor.tar.xz`

Reviewer 3: extra space?

- `snapd_<version>.vendor.tar.xz`

1. Create a GitHub release under https://github.com/canonical/snapd/releases, following the example of previous releases. The `changelog.py` script from the changelog generation step creates a markdown file that can be used here, or you can use NEWS.md and `sed`.
2. Upload the three above-mentioned artifacts to that release
3. Save it as a draft. It should only be published when beta testing has completed. Leave "Set as a pre-release" unchecked.

> #### 10. (Optional but highly recommended) Use a testing PPA to test builds and autopkgtests

Reviewer 3: Would rather flip this round the other way - mandatory but can be skipped in exceptional circumstances
